File: pkg/eventbus/eventbus.go
package eventbus

import (
	"context"
	"sort"
	"sync"
	"unsafe"
)

type (
	Event interface {
		// ResourceType from resource that fired the event
		ResourceType() string

		// EventType returns type of event fired
		EventType() string

		// Match tests if given constraints match
		// event's internal values
		Match(ConstraintMatcher) bool
	}

	eventbus struct {
		// waitgroup for dispatch
		wg *sync.WaitGroup

		// Read & write locking
		// prevent event handling during handler (un)registration
		l *sync.RWMutex

		// list of registered handlers
		handlers map[uintptr]*handler
	}
)

var (
	// Global eventbus
	gEventBus *eventbus
)

func init() {
	gEventBus = New()
}

// Service returns global event bus service
func Service() *eventbus {
	return gEventBus
}

func Set(eb *eventbus) {
	gEventBus = eb
}

func New() *eventbus {
	return &eventbus{
		wg:       &sync.WaitGroup{},
		l:        &sync.RWMutex{},
		handlers: make(map[uintptr]*handler),
	}
}

// WaitFor is synchronous event dispatcher
//
// It waits for each handler and fails on first error
func (b *eventbus) WaitFor(ctx context.Context, ev Event) (err error) {
	b.l.RLock()
	defer b.l.RUnlock()

	for _, t := range b.find(ev) {
		err = func(ctx context.Context, t *handler) error {
			b.wg.Add(1)
			defer b.wg.Done()
			return t.Handle(ctx, ev)

		}(ctx, t)

		if err != nil {
			return
		}
	}

	return
}

// Dispatch runs events asynchronously
func (b *eventbus) Dispatch(ctx context.Context, ev Event) {
	b.l.RLock()
	defer b.l.RUnlock()
	for _, t := range b.find(ev) {
		b.wg.Add(1)
		go func(ctx context.Context, t *handler) {
			defer b.wg.Done()
			_ = t.Handle(ctx, ev)
		}(ctx, t)
	}
}

// Waits for all dispatched events
//
// Should only be used for testing
func (b *eventbus) wait() {
	b.wg.Wait()
}

// Finds all registered handlers compatible with given event
//
// It returns sorted handlers
//
// There is still room for improvement (performance wise) by indexing
// resources and events of each handler.
func (b *eventbus) find(ev Event) (tt HandlerSet) {
	if ev == nil {
		return
	}

	for _, t := range b.handlers {

		if !t.Match(ev) {
			continue
		}

		tt = append(tt, t)
	}

	sort.Sort(tt)

	return
}

// Register creates a new handler with given handler, resource, event with other options and constraints
//
// It returns a handler identifier that can be used to remove (unregister) handler later
func (b *eventbus) Register(h HandlerFn, ops ...HandlerRegOp) uintptr {
	b.l.Lock()
	defer b.l.Unlock()

	var (
		handlers = NewHandler(h, ops...)
		ptr      = uintptr(unsafe.Pointer(handlers))
	)

	b.handlers[ptr] = handlers
	return ptr
}

// Unregister removes one or more registered handlers
func (b *eventbus) Unregister(ptrs ...uintptr) {
	b.l.Lock()
	defer b.l.Unlock()

	for _, ptr := range ptrs {
		delete(b.handlers, ptr)
	}
}

// UnregisterByResource removes one or more registered handlers that match the given resource
func (b *eventbus) UnregisterByResource(r string) {
	b.l.Lock()
	defer b.l.Unlock()
	for p, h := range b.handlers {
		if h.resourceTypes[r] {
			delete(b.handlers, p)
			continue
		}

	}
}

File: lib/js/src/eventbus/eventbus.ts
/**
 * EventBus handles all Corteza events on browser (!! not on corredor server !!)
 *
 * Flow #1
 *  1. Corredor prepares a bundle that is loaded on a client
 *  2. Bundle provides a "callback" function that accepts EventBus object +
 *     all context information and configuration that is needed for
 *     handler registration
 * 3a. When a "Corteza event" is dispatched (Dispatch())
 *     event-bus searches for handler and passes on event information
 * 3b. When manual event is executed (Exec())
 *     event-bus searches for handler and passes on event information
 *
 * Flow #2
 *   1. When web application is initialized, it should register all
 *      explicit server scripts
 *   2. These server scripts are wrapped with a handlerFn that forwards call
 *      to the API (there, request is passed to the Corredor where it's executed)
 *
 */

import { Handler } from './handlers'
import { Event, HandlerFn, onManual, scriptSorter, Trigger } from './shared'

export interface WellKnownPairs {
  [resource: string]: Array<string>;
}

export interface Options {
  pairs: WellKnownPairs;
  strict: boolean;
  verbose: boolean;
}

/**
 * EventBus for event dispatching and handling
 *
 * Since we have much shorter execution path here than we have in case of server scripts,
 * we can afford some optimisation (in comparison to backend's pkg/eventbus)
 */
export class EventBus {
  /**
   * List of wellknown resource & event type pairs
   *
   * If set, eventbus will throw error if unresognized pair is registered or dispatched
   */
  readonly pairs?: WellKnownPairs
  readonly strict: boolean
  readonly verbose: boolean

  private handlers: Handler[] = []

  constructor (opt?: Partial<Options>) {
    this.pairs = opt?.pairs || {}
    this.strict = !!opt?.strict
    this.verbose = !!opt?.verbose
  }

  /**
   * Dispatches event and sequentially calls all handlers.
   *
   * Handling handler results works a bit different then on backend.
   * Scripts executed with handlers have DIRECT access to values passed (by reference)
   * as arguments via event so there's no need to do an explicit return
   *
   * @param {Event} ev Event to dispatch
   */
  async Dispatch (ev: Event, script?: string): Promise<null> {
    if (this.verbose) console.debug('EventBus: event dispatched', { ev, script })

    if (script) {
      if (ev.eventType !== onManual) {
        console.warn('EventBus: explicit events require onManual event type', ev)
        return null
      }
    } else {
      if (ev.eventType === onManual) {
        console.warn('EventBus: implicit events can not define onManual event type', ev)
        return null
      }
    }

    this.checkPairs([ev.resourceType], [ev.eventType])

    const matched = this.find(ev, script)

    if (matched.length === 0) {
      if (this.verbose) console.debug('EventBus: no handlers found', { ev, script, registeredHandlers: this.handlers.length })
      return null
    }

    if (script) {
      // When executing a specific script,
      // make sure we do not run it multiple times.
      matched.splice(1)
    }

    try {
      for (const t of matched) {
        if (this.verbose) console.debug('EventBus: handling event', { ev, trigger: t, script })
        const result = await t.Handle(ev)
        if (result === false) {
          return Promise.reject(new Error('aborted'))
        }
      }
    } catch (err) {
      return Promise.reject(err)
    }

    return null
  }

  /**
   * Filters and sorts all handlers by event & constraints
   */
  private find (ev: Event, script?: string): Handler[] {
    return this.handlers
      .filter(t => t.Match(ev, script))
      .sort(scriptSorter)
  }

  /**
   * Registers Event handler
   *
   * @param handler Handler function
   * @param trigger Trigger definition
   */
  Register (handler: HandlerFn, trigger: Trigger): EventBus {
    if (this.verbose) console.debug('EventBus: event handler registration for', trigger.scriptName, { trigger })

    this.handlers.push(new Handler(handler, trigger))
    return this
  }

  /**
   * Unregisters all handlers
   */
  UnregisterAll (): EventBus {
    this.handlers = []
    return this
  }

  protected checkPairs (resourceTypes: string[], eventTypes: string[]): void {
    if (this.pairs === undefined || !this.strict) {
      return
    }

    resourceTypes.forEach(resourceType => {
      const wket = (this.pairs as WellKnownPairs)[resourceType]
      if (wket === undefined) {
        throw new TypeError('unknown resource type "' + resourceType + "'")
      }

      eventTypes.forEach(eventTypes => {
        if (!wket.includes(eventTypes)) {
          throw new TypeError('unknown event type "' + eventTypes + '" for "' + resourceType + '" resource type')
        }
      })
    })
  }
}
